draft: Kube apply button

deboer-tim opened the pull request:
### What does this PR do?

Draft based on PR #5849, only /renderer files (second commit) are changed here.

Creates an 'Apply YAML' button that prompts for a file, applies it to
the current context, and reports how many resources have been created.

I've added the button to the Deployments, Services, and Ingress/Routes
pages, as these are the most useful pages for a Kubernetes user. I
didn't add it to the Pods page b/c this is less likely used for a
Kube user and it might create confusion with the 'Play Kubernetes YAML'
button.

Not sure about icon - I've used the same icon as the Play button for
now, but could use the full Kube icon or the coming Kube nav icon.

### Screenshot / video of UI

<img width="294" alt="Screenshot 2024-02-07 at 2 46 41 PM" src="https://github.com/containers/podman-desktop/assets/19958075/ebf0208c-2483-4a09-83df-f0ee973716d7">

### What issues does this PR fix or reference?

Fixes #5661.

### How to test this PR?

Pick your favourite yaml and apply!

deboer-tim commented:
Rebased on latest Kube apply branch, updated to use new Kube icon (same as nav), and added tests. Ready for review.

cdrage commented on packages/renderer/src/lib/kube/KubeApplyYAMLButton.svelte:
```diff
@@ -0,0 +1,49 @@
+    await window.showMessageBox({
+      title: 'Kubernetes',
+      type: 'info',
+      message: `${objects.length} resource(s) applied`,
```
Could we say something like: 'Successfully applied {objects.length} resource(s)'?

Just to indicate it's successful :)

deboer-tim replied:
Fixed.

cdrage commented on packages/renderer/src/lib/kube/KubeApplyYAMLButton.svelte:
```diff
@@ -0,0 +1,49 @@
+    await window.showMessageBox({
+      title: 'Kubernetes',
+      type: 'error',
+      message: 'Could not apply Kubernetes yaml: ' + error,
```
```suggestion
      message: 'Could not apply Kubernetes YAML: ' + error,
```

deboer-tim replied:
Fixed.

cdrage reviewed (changes requested):
Success is working, but for some reason not erroring out anything unvalidated / working correctly, and I'm unsure if this is something we didn't check more thoroughly in the other PR:

Working:
https://github.com/containers/podman-desktop/assets/6422176/e22cbdbf-c998-47a0-accf-bae9c396b4c6


When using a random YAML (I just used `helloworld123) in a `test.yaml`):


https://github.com/containers/podman-desktop/assets/6422176/473a9954-e121-4ccc-864b-7fd6eafd2330

deboer-tim commented:
Other comments fixed, and rebased on main.

> Success is working, but for some reason not erroring out anything unvalidated / working correctly, and I'm unsure if this is something we didn't check more thoroughly in the other PR:
> 
> When using a random YAML (I just used `helloworld123) in a `test.yaml`):

I debugged what's happening here: loadManifestsFromFile() loads and parses the file as a JSON object `helloworld123)`, then syncResources() filters it out (b/c it's not a valid manifest with a kind) and doesn't attempt to apply, and returns 0 resources pushed. Even if had tried it would have failed and returned 0 too.

This is kind of an odd case because the caller doesn't know how many valid resource(s) are supposed to be in the file to begin with, i.e. the file could have 5 resources and only 4 were valid/successfully applied, or 3 and 0. I'm thinking there should be two fixes to handle different cases:

1. applyResourcesFromFile() could throw an error if there are no valid resources in the file.
2. The button should special-case 0 and give a warning message that nothing was applied.

I'll do this in my next commit.

deboer-tim commented:
New commit added, as above. Please verify.

cdrage commented:
https://github.com/containers/podman-desktop/assets/6422176/1af30bf3-c834-4c45-b497-8f85a75fcf86

Awesome work! LGTM! Tested via a faulty YAML as well as a valid one.

lstocchi reviewed (commented):
I see this error everything i try to apply a yaml file (no matter if it's valid or not). 
I'm connected to the sandbox cluster.

![image](https://github.com/containers/podman-desktop/assets/49404737/cd3dbf3d-b181-41d5-8967-6d6a88aff306)

cdrage commented:
I've opened up https://github.com/containers/podman-desktop/pull/6086

I can confirm that this PR patched on this one, will result on correct error outputs.

lstocchi reviewed (commented):
The patch made by @cdrage allowed me to read the error.
The problem is that it tries to push the resource to the default namespace even if i'm connected to lstocchi-dev.

![image](https://github.com/containers/podman-desktop/assets/49404737/39763de0-9e86-4b94-a0d9-b85fc3ea8c02)

![image](https://github.com/containers/podman-desktop/assets/49404737/810814db-9a5c-4a33-b5b4-b9707ce5e76c)

deboer-tim commented:
I was going to say the same thing. Maybe you could share the yaml (offline)? If metadata/namespace exists and points to 'default' then this is expected behaviour, otherwise the bug is not directly in this PR but the underlying API (#5849) and I will fix.

lstocchi commented:
This is the yaml
```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: i-17
  annotations:
    nginx.ingress.kubernetes.io/rewrite-target: /
spec:
  ingressClassName: nginx-example
  rules:
  - http:
      paths:
      - path: /testpath1
        pathType: Prefix
        backend:
          resource:
            name: test
            kind: 'sink'
```

deboer-tim commented:
My bad - underlying function supports applying namespace now but I never exposed or hooked it up to the button. Rebased on top of latest to get Charlie's error fix, exposed and added namespace, and updated tests. Working well for me now against sandbox with YAML above.

cdrage commented on packages/renderer/src/lib/kube/KubeApplyYAMLButton.svelte:
```diff
@@ -0,0 +1,59 @@
+  inProgress = true;
+  try {
+    const namespace = await window.kubernetesGetCurrentNamespace();
+    let objects: KubernetesObject[] = await window.kubernetesApplyResourcesFromFile(contextName, file, namespace);
```
🎉

cdrage reviewed (changes requested):
Tested it 3 ways with the following YAML:

```yaml
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: "2024-02-21T21:00:27Z"
  labels:
    app: helloworld-pod
  name: helloworld-pod
spec:
  containers:
  - image: docker.io/cdrage/helloworld:latest
    name: helloworld
    ports:
    - containerPort: 8080
      hostPort: 8080
 ```
 
 
 1. 🎉  `namespace: kube-system` added, even though kubernetes context is `default`. Deploys to `kube-system` which is intentional and awesome.
 2. 🎉 `namespace: default` added, while being on default namespace.
 3. 🛑 No namespace: No Pod is deployed despite it saying "Resources applied". If you apply the same yaml via `kubectl apply` it works fine.

dgolovin commented on packages/main/src/plugin/kubernetes-client.spec.ts:
```diff
@@ -1221,19 +1221,23 @@ test('Expect apply with invalid file should error', async () => {
+    await client.applyResourcesFromFile('default', 'missing-file.yaml');
+    throw Error('should never get here');
+  } catch (err: unknown) {
+    expect(err).to.be.a('Error');
```
Same problem.

dgolovin commented on packages/main/src/plugin/kubernetes-client.spec.ts:
```diff
@@ -1221,19 +1221,23 @@ test('Expect apply with invalid file should error', async () => {
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  } catch (err: any) {
+  } catch (err: unknown) {
     expect(err).to.be.a('Error');
```
```suggestion
```

dgolovin commented on packages/main/src/plugin/kubernetes-client.spec.ts:
```diff
@@ -1221,19 +1221,23 @@ test('Expect apply with invalid file should error', async () => {
+  } catch (err: unknown) {
     expect(err).to.be.a('Error');
-    expect(err.message).equal('File missing-file.yaml does not exist');
+    expect((err as Error).message).equal('File missing-file.yaml does not exist');
```
```suggestion
expectedError = err;
```

dgolovin commented on packages/main/src/plugin/kubernetes-client.spec.ts:
```diff
@@ -1221,19 +1221,23 @@ test('Expect apply with invalid file should error', async () => {
     throw Error('should never get here');
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  } catch (err: any) {
+  } catch (err: unknown) {
```
This test does not fail if apply does not throw an error, but test name says it 'should'.

dgolovin commented on packages/main/src/plugin/kubernetes-client.spec.ts:
```diff
@@ -1221,19 +1221,23 @@ test('Expect apply with invalid file should error', async () => {
   try {
```
```suggestion
  let expectedError: unknown;
  try {
```

dgolovin commented on packages/main/src/plugin/kubernetes-client.spec.ts:
```diff
@@ -1221,19 +1221,23 @@ test('Expect apply with invalid file should error', async () => {
-    expect(err.message).equal('File missing-file.yaml does not exist');
+    expect((err as Error).message).equal('File missing-file.yaml does not exist');
   }
 });
```
```suggestion
  expect(expectedError).to.be.a('Error');
  expect((err as Error).message).equal('File missing-file.yaml does not exist');
});
```

dgolovin commented on packages/renderer/src/lib/kube/KubeApplyYAMLButton.svelte:
```diff
@@ -0,0 +1,59 @@
+    await window.showMessageBox({
+      title: 'Kubernetes',
+      type: 'error',
+      message: 'Could not apply Kubernetes YAML: ' + error,
```
Better to use coercion here
```suggestion
      message: 'Could not apply Kubernetes YAML: ' + String(error),
```

cdrage reviewed (approved):
LGTM! Just updating the tests of @dgolovin 's comments and we should be good to go 🚀 🚀 🚀

deboer-tim commented:
@dgolovin  tests updated, please review

jeffmaury commented:
The Pod page still have Play Kubernetes YAML and not Apply YAML

deboer-tim commented:
> The Pod page still have Play Kubernetes YAML and not Apply YAML

Play Kube is fundamentally different from Apply: it works against Podman (hence the name), or can target Kubernetes but does a create (i.e. won't change existing resources). Apply matches 'kubectl apply' which does 'create or patch', and only works on Kubernetes.

Personally I think we should reduce Play Kube to only work on Podman and use Apply for Kube, but this is not the only area where Pods fall in both camps, having two buttons on the Pods page would be even more confusing, and this gets into some political questions. My goal here was to add Apply Kube to pages like Deployments and Services which are clearly Kube pages and it is expected, and defer/avoid disrupting/wait for feedback on the Pods page.

deboer-tim commented:
Rebased to fix conflicts.

dgolovin reviewed (approved):
LGTM
